source/common/http/codec_client.cc

```diff
@@ -56,6 +46,18 @@ CodecClient::CodecClient(Type type, Network::ClientConnectionPtr&& connection,
 
 CodecClient::~CodecClient() = default;
 
+void CodecClient::connect() {
```

Comment:
It took me a while to figure out why we needed to delay the call to connect().  Some comments may help.

The implementation of CodecClient seems a bit odd since the codec_ member is not initialized until the subclass constructor runs.  I think that some alternate designs to consider would be to pass in the right codec_ as a constructor argument to CodecClient and thus force creation of the QuicHttpClientConnectionImpl before CodecClient's constructor has a chance to call connection_->connect();

Use of composition instead of inheritance for CodecClientProd seems like another potential option but it's hard to say why the current code is written as it is.

Reply:
Another possibly better option to consider is to make connect() a protected method and call it from CodecClientProd's constructor after codec_ is created and the equivalent test class.